lesson1/test/Token.ts

```diff
@@ -0,0 +1,96 @@
+import { expect } from "chai";
+import { ethers } from "hardhat";
+import { SignerWithAddress } from "@nomiclabs/hardhat-ethers/signers";
+import { Contract } from "ethers";
+
+describe("Token", function () {
+  let owner: SignerWithAddress;
+  let otherUser: SignerWithAddress;
+  let token: Contract;
+  const totalSupplyAlsoInitial = "100000000000000000000000";
+  beforeEach(async () => {
+    [owner, otherUser] = await ethers.getSigners();
+    const TokenFactory = await ethers.getContractFactory("Token");
+    token = await TokenFactory.deploy();
+    await token.deployed();
+  });
+
+  describe("token creation", () => {
+    it("Should return token details", async () => {
+      expect(await token.name()).to.equal("Leocode Token");
+      expect(await token.symbol()).to.equal("LEO");
+      expect(await token.decimals()).to.equal(18);
+    });
+
+    it("Should show valid balance of token creator", async () => {
+      expect(await token.balanceOf(owner.address)).to.equal(
+        totalSupplyAlsoInitial
+      );
+    });
+  });
+
+  describe("transfers", () => {
+    it("Should transfer 100 tokens to other user", async () => {
+      await token.transfer(otherUser.address, 100);
```

Comment:
Dzięki trzymaniu wszystkiego na liczbach bardziej czytelnych można potem sobie to jakoś "ogarnąć". Potem sprawdzając np `totalSupply.sub(100)`. Albo przelewając też większe wartości, bo część implementacji wysrywa się czasem przez złą obsługę `decimals` w różnych miejscach